### PR TITLE
Added range argument to farm

### PR DESCRIPTION
**Pull request by millennIumAMbiguity**
farm - farms every crop it can find.
farm 100 - farm crops within a 100 block radius from the starting position.

### CODE REVIEW

**Comment by derBobo** [comment]
is this based on #2054 ?

**Comment by millennIumAMbiguity** [comment]
no, it is based on [#845](https://github.com/cabaletta/baritone/pull/845).

> **feel free to remake the "distance limit arguments to farm" as its own pr tho** - leijurv

**Comment by derBobo** [comment]
maybe you could add the feature to use the home waypoint as startpos, you can copy the code for that from my pr #2088

**Comment by avneh** [comment]
ok i did your code with waypoint so basically two codes in one powerful line

**Comment by l1ving** [comment]
> ok i did your code with waypoint so basically two codes in one powerful line
> https://github.com/cabaletta/baritone/pull/2090/files

I can't tell if you're incredibly stupid or trolling

**Comment by avneh** [comment]
trolling LOL XD

**Comment by avneh** [comment]
sorry i'll get coding

File: src/api/java/baritone/api/process/IFarmProcess.java [hunk 1]
```diff
@@ -17,7 +17,9 @@
 
 package baritone.api.process;
 
+import net.minecraft.util.math.BlockPos;
+
 public interface IFarmProcess extends IBaritoneProcess {
 
-    void farm();
+    void farm(int range, BlockPos pos);
```
Comment (leijurv) [inline src/api/java/baritone/api/process/IFarmProcess.java:24]: Breaks API compatibility, please add a `default`

Comment (leijurv) [reply src/api/java/baritone/api/process/IFarmProcess.java:24]: also would be nice to add a javadoc now that it takes args

File: src/main/java/baritone/command/defaults/FarmCommand.java [hunk 2]
```diff
@@ -34,8 +43,50 @@ public FarmCommand(IBaritone baritone) {
...
+                        throw new CommandInvalidStateException("No waypoints found");
+                    case 1:
+                        waypoint = waypoints[0];
+                        break;
+                    default:
+                        break;
+                }
+            }
+            if (waypoint == null) {
+                throw new CommandInvalidStateException("Multiple waypoints were found");
```
Comment (leijurv) [inline src/main/java/baritone/command/defaults/FarmCommand.java:80]: Move this throw up into the `default:`

File: src/main/java/baritone/command/defaults/FarmCommand.java [hunk 3]
```diff
@@ -34,8 +43,50 @@ public FarmCommand(IBaritone baritone) {
...
+        BetterBlockPos origin;
+        //range
+        if (args.has(1)) {
+            range = args.getAs(Integer.class);
+        }
+        //waypoint
+        if (args.hasExactly(1)){
+            IWaypoint[] waypoints = args.getDatatypeFor(ForWaypoints.INSTANCE);
+            IWaypoint waypoint = null;
+            if (args.hasAny() && args.peekString().equals("@")) {
```
Comment (leijurv) [inline src/main/java/baritone/command/defaults/FarmCommand.java:57]: Is this pasted from somewhere? Should it be a datatype (i don't really know the furry command system that well)?

Just feels like this isn't dry

**Review by leijurv — CHANGES_REQUESTED** [review]

File: src/main/java/baritone/command/defaults/FarmCommand.java (hunk 3 above)
Comment (millennIumAMbiguity) [reply src/main/java/baritone/command/defaults/FarmCommand.java:57]: The waypoints implementation is based on `WaypointsCommand.java` line `152` to `177`.
`args.hasExactly(1)` can be replaced by `args.has(1)` if it's that you're referring to.

File: src/main/java/baritone/command/defaults/FarmCommand.java (hunk 2 above)
Comment (millennIumAMbiguity) [reply src/main/java/baritone/command/defaults/FarmCommand.java:80]: How do you mean?

Comment (ZacSharp) [reply src/main/java/baritone/command/defaults/FarmCommand.java:80]: He means you should put the `throw new CommandInvalidStateException("Multiple waypoints were found");` into the `default:` block and remove the if

File: src/main/java/baritone/command/defaults/FarmCommand.java (hunk 3 above)
Comment (millennIumAMbiguity) [reply src/main/java/baritone/command/defaults/FarmCommand.java]: It was the timestamp compare system, have removed it now.

File: src/main/java/baritone/command/defaults/FarmCommand.java (hunk 2 above)
Comment (millennIumAMbiguity) [reply src/main/java/baritone/command/defaults/FarmCommand.java:67]: I'm unsure of what you mean.

Comment (derBobo) [reply src/main/java/baritone/command/defaults/FarmCommand.java:67]: they mean this
`default:
 throw new CommandInvalidStateException("Multiple waypoints were found");     
break;`

Comment (millennIumAMbiguity) [reply src/main/java/baritone/command/defaults/FarmCommand.java:67]: Sorry, was not refreshing the page and didn't see any replays. I thought you meant the interface default.

**Pull request merged**